### PR TITLE
Adds elixir support, when mix is accessible

### PR DESCRIPTION
**Pull request by meghfossa**
# Overview

This PR adds elementary elixir support using mix. I wanted to keep this PR small and manageable, so would like to merge in first tactic, before adding in tactic for parsing `mix.lock` and `mix.exs` file for analysis.

## Acceptance criteria

- [x] CLI can infer direct and indirect dependencies when `mix` is accessible 
- [x] CLI can analyze dependency from git, or hex when `mix` is accessible 

## Testing plan

- I tested this locally against few open sources elixir project. I performed e2e test against results generated. 

To test, 

- run .`/fossa analyze projectDir --output` for example project and compare result against `mix deps.tree` and `mix deps`

## Risks

_Highlight any areas that you're unsure of, or want reviewers to pay particular attention to._

Mix supports multiple build managers - mix, rebar, rebar3, and make. Collected dependency could be discovered as its own project from rebar, rebar3 strategies. This may lead to noise when running the analyze command, and in the results. Currently, there is no way to filter out specific directory. 

## References

- https://github.com/fossas/team-analysis/issues/398
- https://github.com/fossas/team-analysis/issues/661
- https://github.com/fossas/team-analysis/issues/660

This does not close, but partially meets reqs of https://github.com/fossas/team-analysis/issues/398

## Checklist

- [x] I added tests for this PR's change (or confirmed tests are not viable).
- [x] I linted and formatted (via `haskell-language-server`) any haskell files I touched in this PR.
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I linked this PR to any referenced GitHub issues, if they exist.

### CODE REVIEW

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [inline src/Strategy/Elixir/MixTree.hs:326]: I would like to abstract this out - it's very similar to poetry's version constraint expression parsing. But keeping this in for now, until poetry PR is merged. I can address this in separate PR.

**Comment by zlav** [comment]
Can you elaborate on what this means?

>Collected dependency could be discovered as its own project from rebar, rebar3 strategies

Are you saying that there vendored dependencies that have a `rebar.config` file which will be picked up and scanned?

**Comment by meghfossa** [comment]
> Are you saying that there vendored dependencies that have a `rebar.config` file which will be picked up and scanned?

Ah! As mix can use few build manager (make, rebar, rebar3, mix). Some dependency may, for example have rebar3 as build manager. 

Here is an example:

mix.lock file
```elixir
%{
...

  "hackney": {:hex, :hackney, "1.15.2", "07e33c794f8f8964ee86cebec1a8ed88db5070e52e904b8f12209773c1036085", [:rebar3], [{:certifi, "2.5.1", [hex: :certifi, repo: "hexpm", optional: false]}, {:idna, "6.0.0", [hex: :idna, repo: "hexpm", optional: false]}, {:metrics, "1.0.1", [hex: :metrics, repo: "hexpm", optional: false]}, {:mimerl, "~>1.1", [hex: :mimerl, repo: "hexpm", optional: false]}, {:ssl_verify_fun, "1.1.5", [hex: :ssl_verify_fun, repo: "hexpm", optional: false]}], "hexpm", "e0100f8ef7d1124222c11ad362c857d3df7cb5f4204054f9f0f4a728666591fc"},

...
}```

When compiling and getting dependencies, it will download build config for `hackney` in deps directory.

```
>> tree elxirEampleProject/deps/hackney
...
├── _build
│   └── default
│       └── lib
├── ebin
├── hex_metadata.config
├── rebar.config
├── rebar.lock
└── src
    ├── hackney.app.src
....
```

This is getting discovered as it's own project with analyze. I'm not sure how to exclude this. I did skip deps directory in my strategy's discovery, but since rebar strategies are independent they pick these up as independent projects.

**Comment by scruffystuffs** [comment]
The way to solve this is to adjust the rebar strategy to not search the `deps` folder if there's a `mix.exs` file with the string `":rebar3"`.

**Review by zlav — COMMENTED** [review]
This looks really good. I have a few comments that need to be addressed and a few nits but nothing that is too major.

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs:110]: No you are right. By default command reports all dependencies, I can update it so we only get prod dependencies.

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:75]: I'd rather not see `<&>` used in the codebase, we don't really use it or `&` anywhere else.
Also, the `<$>` in `(supportedSCM <$> depResolvedSCM)` is just `.`.  I mentioned it on the poetry PR, but `-> a` has a functor instance, where `fmap = (.)`.  I've corrected that in my suggestion.

```suggestion
  depsAllEnvTree <- 
    context "Identifying relationship among dependencies" $
      supportedMixDeps <$> execParser mixTreeCmdOutputParser dir mixDepTreeCmd
  depsAllResolved <-
    context "Inferring dependencies versioning" $
      Map.filter (supportedSCM . depResolvedSCM)
        execParser mixDepsCmdOutputParser dir mixDepCmd
```

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:196]: [nit]: Use `toText` from `Data.String.Conversion` instead of `Text.pack`.
```suggestion
    findVersion = toText <$> some (alphaNumChar <|> char '.' <|> char '-')
```

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:211]: We talked offline about using `space` directly, but I just wanted to leave a comment so we didn't forget: use a local `sc` or `scn` (**S**pace **C**onsumer, or **S**pace **C**onsumer with **N**ewlines), see other strategies for examples

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:241]: [nit]: use `void` from Data.Functor` instead.
```suggestion
    ignored = void $ takeWhileP (Just "ignored") (not . isEndLine)
```

Comment (scruffystuffs) [reply src/Strategy/Elixir/MixTree.hs:241]: from the source for `Data.Functor`

```haskell
void :: Functor f => f a -> f ()
void x = () <$ x
```

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:265]: [optional nit]: prefer `>>=` to `=<<`.  It's a little unintuitive considering that most similar operators go the other direction, but it does seem to be more idiomatic in other haskell libraries.
```suggestion
      pure [MixDep (PackageName dep) (version >>= toDependencyVersion) scm (concat deps)]
```

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:377]: [nit]: prefer `toString` from `Data.String.Conversion` over `Text.unpack`
```suggestion
      other -> fail ("Could not recognize mix constraint operator: (" <> toString other <> ")")
```

File: docs/strategies/elixir.md
Comment (scruffystuffs) [inline docs/strategies/elixir.md:23]: If we've changed our commands to handle prod vs. dev differently, we should sync these docs with those changes.  if not, then these are OK.

File: src/Strategy/Mix.hs
Comment (scruffystuffs) [inline src/Strategy/Mix.hs:1]: Love the size and organization of this file 👍

**Review by scruffystuffs — CHANGES_REQUESTED** [review]
A few changes, but along with the feedback from @zlav, this is pretty close.

File: docs/strategies/elixir.md
Comment (meghfossa) [reply docs/strategies/elixir.md]: No we need to explicitly say `--all`. If not done so, environment variable `MIX_ENV` can influence results returned. This looks right to me.

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs:388]: I've added this suggestion to style guide now - https://github.com/fossas/team-analysis/issues/671

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs]: +in style guide: https://github.com/fossas/team-analysis/issues/671

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs:229]: +style guide https://github.com/fossas/team-analysis/issues/671

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs]: +style guide https://github.com/fossas/team-analysis/issues/671

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs]: +style guide https://github.com/fossas/team-analysis/issues/671

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs:354]: I've modified the code a bit to collapse multiple function into one. I intent to abstract this piece of code out in future PR, once poetry PR is merged. The parsing for version constraint is fairly similar among them - and I'm presuming it will be for lots of languages. I tried the above code, but it led into test failing for irregular spacing. If you insist I can tackle it in this pr. My preference is that I address this subsequent PR.

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs]: I've renamed to Other. I would like to keep this as a catch all - I've updated a code to filter out Others when resolving graph dependencies - that I think should take care of it.

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs]: Yeah - I don't think it's possible, unless new operators are added to operator list without updating textToMixVersion.

File: docs/strategies/elixir.md
Comment (scruffystuffs) [reply docs/strategies/elixir.md]: It looks like the first command changed to include prod, right?  you added `--only prod`

Comment (meghfossa) [reply docs/strategies/elixir.md]: That's correct. I want to capture all resolved dependencies with `mix deps --all`. 

The code only reports dependencies from `mix deps.tree --only prod` but uses resolved versions captured from `mix deps --all`.

File: docs/strategies/elixir.md
Comment (scruffystuffs) [inline docs/strategies/elixir.md]:
```suggestion
1. Run `mix deps.tree --format plain --only prod` and generate output similar to:
```

File: docs/strategies/elixir.md
Comment (scruffystuffs) [reply docs/strategies/elixir.md]: I might not be as clear as I think I'm being.  See [this comment](https://github.com/fossas/spectrometer/pull/287/files/99cea438c20e58879ceb761d421f18f47b4e658b#r674192081) for what I was expecting.

File: docs/strategies/elixir.md
Comment (scruffystuffs) [reply docs/strategies/elixir.md]: nvm, this is already done, we were discussing an old commit after you had fixed it, my bad.

File: docs/strategies/elixir.md
Comment (scruffystuffs) [reply docs/strategies/elixir.md]: I just noticed that this is done in the latest commit, and we're commenting on an old commit.

File: src/Strategy/Elixir/MixTree.hs
Comment (meghfossa) [reply src/Strategy/Elixir/MixTree.hs:343]: I was able to make this work. This should have been `Nothing -> return $ CEq versionText` instead of `Nothing -> fail ("Could not recognize mix constraint operator:")`. This is for case when user simply does not use any operator e.g. myPkg: "2.2" vs myPkg: ">2.2"

**Comment by meghfossa** [comment]
> > Are you saying that there vendored dependencies that have a `rebar.config` file which will be picked up and scanned?
> 
> Ah! As mix can use few build manager (make, rebar, rebar3, mix). Some dependency may, for example have rebar3 as build manager.
> 
> Here is an example:
> 
> mix.lock file
> 
> ```elixir
> %{
>   "hackney": {:hex, :hackney, "1.15.2", "07e33c794f8f8964ee86cebec1a8ed88db5070e52e904b8f12209773c1036085", [:rebar3], [{:certifi, "2.5.1", [hex: :certifi, repo: "hexpm", optional: false]}, {:idna, "6.0.0", [hex: :idna, repo: "hexpm", optional: false]}, {:metrics, "1.0.1", [hex: :metrics, repo: "hexpm", optional: false]}, {:mimerl, "~>1.1", [hex: :mimerl, repo: "hexpm", optional: false]}, {:ssl_verify_fun, "1.1.5", [hex: :ssl_verify_fun, repo: "hexpm", optional: false]}], "hexpm", "e0100f8ef7d1124222c11ad362c857d3df7cb5f4204054f9f0f4a728666591fc"},
> }
> ```
> 
> When compiling and getting dependencies, it will download build config for `hackney` in deps directory.
> 
> ```
> >> tree elxirEampleProject/deps/hackney
> ...
> ├── _build
> │   └── default
> │       └── lib
> ├── ebin
> ├── hex_metadata.config
> ├── rebar.config
> ├── rebar.lock
> └── src
>     ├── hackney.app.src
> ....
> ```
> 
> This is getting discovered as it's own project with analyze. I'm not sure how to exclude this. I did skip deps directory in my strategy's discovery, but since rebar strategies are independent they pick these up as projects.

I and @zlav discussed potential options in regards to this. Options discussed were, 

1) Modify Rebar3 strategy to exlude _deps directory when mix.exs is found. 
2) No Code Change. Use filter options to exclude _deps directory. 

We decided to go with (2) for now. Main reason for doing so is that approach (1) introduces a change in sibling strategy (2) It deviates the notion of independent strategies and introduces some coupling. Aforementioned concern may only be applicable to small segment of elixir project. For such project alternate solution exists as of now. I will update relevant docs for the strategy.

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:354]: [nit]: Remember to use `pure` instead of `return`

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:365]: [nit]: Use `toText` from `Data.String.Conversion`
```suggestion
    findVersionText = toText <$> some (alphaNumChar <|> char '.' <|> char '-' <|> char '*' <|> char '+')
```

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:360]: [nit]: Use `$` and `.` instead of brackets in easy cases like this.
```suggestion
    parseConstraintOperator = optional . asum $ map symbol operatorList
```

**Review by scruffystuffs — COMMENTED** [review]
Few nits, I'll do a full review soon.

File: src/Strategy/Elixir/MixTree.hs
Comment (scruffystuffs) [inline src/Strategy/Elixir/MixTree.hs:183]: [nit]: Avoid `do`-notation
```suggestion
    parseDepHex = Hex <$ chunk "Hex package"
```

**Review by scruffystuffs — APPROVED** [review]
Looks good to me.  You'll need to rebase on master to get CI to pass.  I left a few nits, but nothing else needs to be addressed, you can merge this when you're ready.

Make sure you fix the changelog before merging.

**Comment by meghfossa** [comment]
> > I and @zlav discussed potential options in regards to this. Options discussed were,
> > 
> > 1. Modify Rebar3 strategy to exlude _deps directory when mix.exs is found.
> > 2. No Code Change. Use filter options to exclude _deps directory.
> > 
> > We decided to go with (2) for now. Main reason for doing so is that approach (1) introduces a change in sibling strategy (2) It deviates the notion of independent strategies and introduces some coupling. Aforementioned concern may only be applicable to small segment of elixir project. For such project alternate solution exists as of now. I will update relevant docs for the strategy.
> 
> The reason I suggested doing number 1 is that it's really easy to do, usually correct (as long as it's fully checked), and already has precedent in our codebase ([`npm` ignores `package.json` if it also finds `yarn.lock`](https://github.com/fossas/spectrometer/blob/b436f4fab8da37da1b1066f22b67d24932c53751/src/Strategy/Npm.hs#L24)). While most good code is dumb code, scanning for and identifying projects is one of the few things that needs more nuanced processing. We don't have to do it in this PR, but I'd like to handle it sooner rather than later.
> 
> We can't keep strategies truly independent if they don't actually stay independent in the real world. Coupling in our codebase is unavoidable if the actual tools are coupled.

Only potential issue I see with this approach is that it will be incorrect, if the projects follow umbrella structure. Albeit, this may be not as frequent.

**Pull request merged**
